podbuilder/podbuilder.go

```diff
@@ -0,0 +1,351 @@
+/*
+Copyright 2020 gRPC authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podbuilder
+
+import (
+	"fmt"
+
+	"github.com/grpc/test-infra/kubehelpers"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	grpcv1 "github.com/grpc/test-infra/api/v1"
+	"github.com/grpc/test-infra/config"
+)
+
+// addReadyInitContainer configures a ready init container. This container is
+// meant to wait for workers to become ready, writing the IP address and port of
+// these workers to a file. This file is then shared over a volume with the
+// driver's run container.
+//
+// This method also sets the $QPS_WORKERS_FILE environment variable on the
+// driver's run container. Its value will point to the aforementioned, shared
+// file.
+func addReadyInitContainer(defs *config.Defaults, test *grpcv1.LoadTest, podspec *corev1.PodSpec, container *corev1.Container) {
+	if defs == nil || podspec == nil || container == nil {
+		return
+	}
+
+	readyContainer := newReadyContainer(defs, test)
+	podspec.InitContainers = append(podspec.InitContainers, readyContainer)
+
+	container.Env = append(container.Env, corev1.EnvVar{
+		Name:  "QPS_WORKERS_FILE",
+		Value: config.ReadyOutputFile,
+	})
+
+	container.VolumeMounts = append(container.VolumeMounts, corev1.VolumeMount{
+		Name:      config.ReadyVolumeName,
+		MountPath: config.ReadyMountPath,
+	})
+
+	podspec.Volumes = append(podspec.Volumes, corev1.Volume{
+		Name: config.ReadyVolumeName,
+	})
+}
+
+// newReadyContainer constructs a container using the default ready container
+// image. If defaults parameter is nil, an empty container is returned.
+func newReadyContainer(defs *config.Defaults, test *grpcv1.LoadTest) corev1.Container {
+	if defs == nil {
+		return corev1.Container{}
+	}
+
+	var args []string
+	for _, server := range test.Spec.Servers {
+		args = append(args, fmt.Sprintf("%s=%s,%s=%s,%s=%s",
+			config.LoadTestLabel, test.Name,
+			config.RoleLabel, config.ServerRole,
+			config.ComponentNameLabel, *server.Name,
```

Comment:
Nit: Wondering if we are trying to eliminate the component object why are we still using `componentNameLabel` ? If it means different from what it was you might want to change the comment for `componentNameLabel` in constant.go.

Reply:
Agreed. Will figure out a rename or reword.

Reply:
@wanlin31 I'm not sure of a better name for this. I don't want to call it the pod name b/c it is not. I think drivers, servers and clients are still components of the test. So, we may just want to leave as is for the moment and revisit if we have time later.